src/difference.jl

```diff
@@ -0,0 +1,45 @@
+"""
+    difference(ε::Real, y::T, x::T) where {T}
+
+Computes `dx` where `dx` is defined s.t.
+```julia
+y = x + ε * dx
+```
+where `dx` is a valid tangent type for `x`.
+
+If `(y - x) / ε` is defined, then this operation is equivalent to doing that. For functions
+where these operations aren't defined, `difference` can still be defined without commiting
+type piracy while `-` and `/` cannot.
+"""
+difference(::Real, ::T, ::T) where {T<:Symbol} = DoesNotExist()
+difference(::Real, ::T, ::T) where {T<:AbstractChar} = DoesNotExist()
+difference(::Real, ::T, ::T) where {T<:AbstractString} = DoesNotExist()
+difference(::Real, ::T, ::T) where {T<:Integer} = DoesNotExist()
+
+difference(ε::Real, y::T, x::T) where {T<:Number} = (y - x) / ε
+
+difference(ε::Real, y::T, x::T) where {T<:StridedArray} = difference.(ε, y, x)
```

Comment:
why does this need to be restricted to `StridedArray` rather than `AbstractArray`?

Reply:
So that things like `Diagonal`s etc are handled by the generic struct interface, so as to prevent the creation of dense tangents for e.g. `Diagonal`s.

Reply:
https://github.com/JuliaDiff/ChainRules.jl/issues/232